File: Backend/Capstone2/src/main/java/C2SE/_1/Capstone2/service/ShiftService.java
package C2SE._1.Capstone2.service;

import C2SE._1.Capstone2.dto.ShiftAssignmentCreateDTO;
import C2SE._1.Capstone2.dto.ShiftAssignmentDTO;
import C2SE._1.Capstone2.dto.ShiftAssignmentUpdateDTO;
import C2SE._1.Capstone2.dto.ShiftAttendanceDTO;
import C2SE._1.Capstone2.dto.ShiftBulkAssignDTO;
import C2SE._1.Capstone2.dto.ShiftCheckInDTO;
import C2SE._1.Capstone2.dto.ShiftCheckOutDTO;
import C2SE._1.Capstone2.dto.ShiftRevenueDetailDTO;
import C2SE._1.Capstone2.dto.ShiftTemplateDTO;
import C2SE._1.Capstone2.dto.ShiftWorkSummaryDTO;
import C2SE._1.Capstone2.entity.ShiftType;

import java.time.LocalDate;
import java.util.List;

public interface ShiftService {

    List<ShiftTemplateDTO> getShiftTemplates(boolean activeOnly, ShiftType shiftType);

    ShiftTemplateDTO createShiftTemplate(ShiftTemplateDTO dto);

    ShiftTemplateDTO updateShiftTemplate(Long id, ShiftTemplateDTO dto);

    void deactivateShiftTemplate(Long id);

    List<ShiftAssignmentDTO> getAssignments(LocalDate fromDate, LocalDate toDate, Long userId, ShiftType shiftType);

    List<ShiftAssignmentDTO> getMyAssignments(LocalDate fromDate, LocalDate toDate);

    ShiftAssignmentDTO createAssignment(ShiftAssignmentCreateDTO dto);

    List<ShiftAssignmentDTO> createAssignmentsBulk(ShiftBulkAssignDTO dto);

    ShiftAssignmentDTO updateAssignment(Long id, ShiftAssignmentUpdateDTO dto);

    ShiftAssignmentDTO cancelAssignment(Long id);

    ShiftAttendanceDTO checkIn(ShiftCheckInDTO dto);

    ShiftAttendanceDTO checkOut(ShiftCheckOutDTO dto);

    List<ShiftAttendanceDTO> getAttendances(LocalDate fromDate, LocalDate toDate, Long userId, ShiftType shiftType);

    List<ShiftAttendanceDTO> getMyAttendances(LocalDate fromDate, LocalDate toDate);

    List<ShiftWorkSummaryDTO> getWorkSummary(LocalDate fromDate, LocalDate toDate, Long userId, ShiftType shiftType);

    List<ShiftRevenueDetailDTO> getRevenueDetails(LocalDate fromDate, LocalDate toDate, Long userId, ShiftType shiftType);
}

File: Frontend/smartdss-frontend/src/services/shiftService.ts
import api from './api';
import type {
  ApiResponse,
  ShiftAssignment,
  ShiftAssignmentCreatePayload,
  ShiftAssignmentUpdatePayload,
  ShiftAttendance,
  ShiftBulkAssignPayload,
  ShiftCheckPayload,
  ShiftRevenueDetail,
  ShiftTemplate,
  ShiftTemplateForm,
  ShiftType,
  ShiftWorkSummary,
} from '@/types';

export const shiftService = {
  getTemplates: (activeOnly = true, shiftType?: ShiftType) =>
    api.get<ApiResponse<ShiftTemplate[]>>('/shifts/templates', { params: { activeOnly, shiftType } }),

  createTemplate: (data: ShiftTemplateForm) =>
    api.post<ApiResponse<ShiftTemplate>>('/shifts/templates', data),

  updateTemplate: (id: number, data: ShiftTemplateForm) =>
    api.put<ApiResponse<ShiftTemplate>>(`/shifts/templates/${id}`, data),

  deactivateTemplate: (id: number) =>
    api.delete<ApiResponse<void>>(`/shifts/templates/${id}`),

  getAssignments: (params?: { fromDate?: string; toDate?: string; userId?: number; shiftType?: ShiftType }) =>
    api.get<ApiResponse<ShiftAssignment[]>>('/shifts/assignments', { params }),

  getMyAssignments: (params?: { fromDate?: string; toDate?: string }) =>
    api.get<ApiResponse<ShiftAssignment[]>>('/shifts/my-assignments', { params }),

  createAssignment: (data: ShiftAssignmentCreatePayload) =>
    api.post<ApiResponse<ShiftAssignment>>('/shifts/assignments', data),

  createAssignmentsBulk: (data: ShiftBulkAssignPayload) =>
    api.post<ApiResponse<ShiftAssignment[]>>('/shifts/assignments/bulk', data),

  updateAssignment: (id: number, data: ShiftAssignmentUpdatePayload) =>
    api.put<ApiResponse<ShiftAssignment>>(`/shifts/assignments/${id}`, data),

  cancelAssignment: (id: number) =>
    api.delete<ApiResponse<ShiftAssignment>>(`/shifts/assignments/${id}`),

  checkIn: (data: ShiftCheckPayload) =>
    api.post<ApiResponse<ShiftAttendance>>('/shifts/attendance/check-in', data),

  checkOut: (data: ShiftCheckPayload) =>
    api.post<ApiResponse<ShiftAttendance>>('/shifts/attendance/check-out', data),

  getAttendances: (params?: { fromDate?: string; toDate?: string; userId?: number; shiftType?: ShiftType }) =>
    api.get<ApiResponse<ShiftAttendance[]>>('/shifts/attendance', { params }),

  getMyAttendances: (params?: { fromDate?: string; toDate?: string }) =>
    api.get<ApiResponse<ShiftAttendance[]>>('/shifts/my-attendance', { params }),

  getWorkSummary: (params?: { fromDate?: string; toDate?: string; userId?: number; shiftType?: ShiftType }) =>
    api.get<ApiResponse<ShiftWorkSummary[]>>('/shifts/work-summary', { params }),

  getRevenueDetails: (params?: { fromDate?: string; toDate?: string; userId?: number; shiftType?: ShiftType }) =>
    api.get<ApiResponse<ShiftRevenueDetail[]>>('/shifts/revenue-details', { params }),

  exportWorkSummaryCsv: (params?: { fromDate?: string; toDate?: string; userId?: number; shiftType?: ShiftType }) =>
    api.get<Blob>('/shifts/work-summary.csv', {
      params,
      responseType: 'blob',
    }),
};
